### ISSUE TITLE
Detect virtual threads and reduce memory needs

### ISSUE THREAD

**Issue by StefanLobbenmeierObjego** [paketo-buildpacks/spring-boot#419]
## Describe the Enhancement

Virtual Threads require less memory ([source](https://spring.io/blog/2022/10/11/embracing-virtual-threads#where-virtual-threads-make-sense)). Now that spring boot supports them in 3.2, is it possible to detect that they are enabled and reduce thread memory in the build pack?

## Possible Solution

There is already detection for reactive spring, which reduces the thread count:

```
If the application is a reactive web application
Configures $BPL_JVM_THREAD_COUNT to 50
```

Since virtual threads are configured in the application.properties, I am not sure if it is also possible or if the build pack just looks at dependencies.

## Motivation

Less memory reserved for threads = more memory for the application, or cheaper hosting costs.

Also looking for input from others, because I am not sure how much memory this will save / How many threads are still there with this setting. For me the best way to test this would be to run the application once with virtual threads and once without and compare memory needs.

**Comment by dmikusa** [comment]
I really like this idea. It would be great if buildpacks could auto-detect and adapt to this difference in configuration. That would be a nice benefit of using buildpacks.

The way that we do this now for Reactive/Servlet/Non-web is to look at what classes are present in the application. If there were a way to determine if virtual threads were being used in this manner, that would be an option. Because virtual threads are part of the JVM and not a JAR that gets included with the app, I'm not sure this option would work. I'm not an expert on virtual threads either though, if someone knows of a way we could make this work let me know.

I'm not 100% sure how we might do that in a manageable way. Because of the complexity of `application.properties`/`application.yml`, attempting to read values out of those files is not something that I think we can reliably do. We at least try to steer clear of this approach. What we might be able to do is to look for system properties or environment variables that would indicate virtual threads are being used. That would not cover the case where someone enabled it directly in application.properties|yml, but if `-Dspring.threads.virtual.enabled=true` is set or if `SPRING_THREADS_VIRTUAL_ENABLED=true`, we could easily detect that, and adjust the configuration. It's not perfect, but may be better than nothing.

Another way that might be an option would be to flip things around and have the user declare to buildpacks that they want virtual threads enabled. Then the buildpack could set the corresponding env variable or system property to enable it and at the same time reconfigure the number of threads set. I'm not a huge fan of this approach though as I like to avoid introducing new buildpack-specific settings like this. It increases the complexity of buildpacks and what users have to know to make them work. I'm a much bigger fan of making things "just work" based off existing settings and options users already know.

I'm not also totally sure how users will want to configure virtual threads, and I think that kind of matters. If it's something users will want to enable across all environments (i.e. dev/test/prod) or if they might want to turn it on/off here and there. The former is probably easier done via application.properties, while the latter is better with external configuration. The way people use it might affect the approach we want to take as well.

Anyway, this is all me thinking out loud. I really like the suggestion, if folks have thoughts on the implementation feel free to post them here. We're listening.

**Comment by StefanLobbenmeierObjego** [comment]
Yeah I figured that parsing the application.yaml is probably a bit too difficult, especially considering that the files might be inactive depending on profile. So I agree with that it is probably easier to flip this around as you said. We want to roll out virtual threads in the coming weeks, so maybe I can also post my insights on how memory usage changes here.

**Comment by dmikusa** [comment]
> We want to roll out virtual threads in the coming weeks, so maybe I can also post my insights on how memory usage changes here.

That would be fantastic! Thanks

**Comment by burl21** [comment]
Any updates?

**Comment by balazspeterfi** [comment]
Hey guys, any update on this? One of the main reasons for me to switch from Google's Jib to Paketo Buildpacks is this cool memory calculator... but the current version doesn't allow any java app to start below 500Mb of container memory partially due to the default thread count used for Tomcat, which is totally wrong when virtual threads are enabled. For sure this can be overridden, but the whole point of this tool is to cleverly figure it out itself.

**Comment by dmikusa** [comment]
> Hey guys, any update on this? One of the main reasons for me to switch from Google's Jib to Paketo Buildpacks is this cool memory calculator... but the current version doesn't allow any java app to start below 500Mb of container memory partially due to the default thread count used for Tomcat, which is totally wrong when virtual threads are enabled. 

The main blocker is how we detect this. It wouldn't be hard to introduce an env variable for buildpacks that you can set to indicate you're using virtual threads, but then there's not a huge advantage of just setting the thread count, which you can already do.

As I mentioned above, reading application.yml|properties with perfect resolution isn't really feasible. The process is fairly complicated, with profiles and env variables and system properties and all the different sources Spring takes into account. It's likely that we would miss things if we attempted this. That said, if folks could provide feedback on how they are enabling virtual threads in their application, and we see some trends then perhaps we could read this is a constrained way like only reading the literal application.properties or application.yml (and no other sources).

I'll get things started. Personally, the apps I have that are using virtual threads all enable it in application.yml, and it's always enabled. I'm not using profiles to conditionally enable it. I'm not enabling it via env variable or system property.

How are you all enabling virtual threads?

> For sure this can be overridden, but the whole point of this tool is to cleverly figure it out itself.

If you want to do this, you'd set `BPL_JVM_THREAD_COUNT` at runtime. So `docker run -e BPL_JVM_THREAD_COUNT=50 ...` 

I don't know what a reasonable value would be. We use 50 for reactive apps, so that's probably reasonable here too. There will still be platform threads even when using virtual threads, I'm just not sure how many you'd typically see. If you're using virtual threads, please take a moment to look at your apps and see how many platform threads are running and report back. This will help pick a good default.

**Comment by balazspeterfi** [comment]
> The main blocker is how we detect this. It wouldn't be hard to introduce an env variable for buildpacks that you can set to indicate you're using virtual threads, but then there's not a huge advantage of just setting the thread count, which you can already do.

I agree that if it's just another environment variable it doesn't have much added value over using `BPL_JVM_THREAD_COUNT` (which is quite confusing to me when set from Maven). Ideally it should be picked up automatically which I understand is more than challenging.

> How are you all enabling virtual threads?

We always do it through `application.yml` too. Although possible, I don't see the point of it being profile dependent as we want to run the same setup on all our environments be it dev/test/prod.
